In [116]:
import pandas as pd


In [117]:
df=pd.read_csv("/content/train.csv")
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [118]:
df["Age"]=df["Age"].fillna(df["Age"].median())
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [119]:
df["Emabrked"]=df["Embarked"].fillna(df["Embarked"].mode())
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Emabrked  
0        0         A/5 21171   7.2500   NaN        S        S  
1        0          PC 17599  71.2833   C85        C        C  
2        0  STON/O2. 3101282   7.9250   NaN        S        S  
3        0            113803  53.1000  C123        S        S  
4        0            373450   8.0500   NaN        S        S  
..     ...               ...      ...   ...      ...      ...  
886      0            211536  13.0000   NaN        S        S  
887      0            112053  30.0000   B42        S        S  
888      2        W./C. 6607  23.4500   NaN        S        S  
889      0            111369  30.0000  C148        C        C  
890      0            370376   7.7500   NaN        Q        Q  

[891 rows x 13 columns]

In [120]:
x=df.drop("Survived",axis=1)
y=df["Survived"]
print(x)
print(y)

     PassengerId  Pclass                                               Name  \
0              1       3                            Braund, Mr. Owen Harris   
1              2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3       3                             Heikkinen, Miss. Laina   
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5       3                           Allen, Mr. William Henry   
..           ...     ...                                                ...   
886          887       2                              Montvila, Rev. Juozas   
887          888       1                       Graham, Miss. Margaret Edith   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   
889          890       1                              Behr, Mr. Karl Howell   
890          891       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket  

In [121]:
"""x=x.drop(["PassengerId"],axis=1)
x.shape"""

'x=x.drop(["PassengerId"],axis=1)\nx.shape'

In [122]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Pclass       891 non-null    int64  
 2   Name         891 non-null    object 
 3   Sex          891 non-null    object 
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Cabin        204 non-null    object 
 10  Embarked     889 non-null    object 
 11  Emabrked     889 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 83.7+ KB


In [123]:
df["Sex"].value_counts(dropna=False)

Sex
male      577
female    314
Name: count, dtype: int64

In [124]:
"""df["Sex"]=df["Sex"].astype(str)
df["Sex"]=df["Sex"].str.strip().str.lower()
df["Sex"]=df["Sex"].map({"male":0, "female":1})
df"""

'df["Sex"]=df["Sex"].astype(str)\ndf["Sex"]=df["Sex"].str.strip().str.lower()\ndf["Sex"]=df["Sex"].map({"male":0, "female":1})\ndf'

In [125]:
df["Sex"].value_counts(dropna=False)


Sex
male      577
female    314
Name: count, dtype: int64

In [126]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Emabrked  
0      0         A/5 21171   7.2500   NaN        S        S  
1      0          PC 17599  71.2833   C85        C        C  
2      0  STON/O2. 3101282   7.9250   NaN        S        S  
3      0            113803  53.1000  C123        S        S  
4      0            373450   8.0500   NaN        S        S

In [127]:
# Encode Sex
df["Sex"] = df["Sex"].map({"male": 0, "female": 1})

# One-hot encode Embarked
df = pd.get_dummies(df, columns=["Embarked"], drop_first=True)

# Drop useless columns
df = df.drop(columns=["Name", "Ticket"], errors="ignore")


In [128]:
df["Sex"].value_counts(dropna=False)

Sex
0    577
1    314
Name: count, dtype: int64

In [129]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Pclass       891 non-null    int64  
 2   Name         891 non-null    object 
 3   Sex          891 non-null    object 
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Cabin        204 non-null    object 
 10  Embarked     889 non-null    object 
 11  Emabrked     889 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 83.7+ KB


In [150]:
#Train-Test Split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42)

In [151]:
x_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 623 entries, 445 to 102
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  623 non-null    int64  
 1   Pclass       623 non-null    int64  
 2   Name         623 non-null    object 
 3   Sex          623 non-null    object 
 4   Age          623 non-null    float64
 5   SibSp        623 non-null    int64  
 6   Parch        623 non-null    int64  
 7   Ticket       623 non-null    object 
 8   Fare         623 non-null    float64
 9   Cabin        139 non-null    object 
 10  Embarked     622 non-null    object 
 11  Emabrked     622 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 63.3+ KB


In [152]:
from sklearn.linear_model import LogisticRegression
import pandas as pd

x_train['Sex'] = x_train['Sex'].map({"male": 0, "female": 1})
x_test['Sex'] = x_test['Sex'].map({"male": 0, "female": 1})

x_train = x_train.drop(columns=["PassengerId", "Name", "Ticket"], errors='ignore')
x_test = x_test.drop(columns=["PassengerId", "Name", "Ticket"], errors='ignore')


x_train = pd.get_dummies(x_train, columns=["Embarked"], drop_first=True)
x_test = pd.get_dummies(x_test, columns=["Embarked"], drop_first=True)

train_cols = x_train.columns
test_cols = x_test.columns

missing_in_test = set(train_cols) - set(test_cols)
for c in missing_in_test:
    x_test[c] = 0

missing_in_train = set(test_cols) - set(train_cols)
for c in missing_in_train:
    x_train[c] = 0

x_test = x_test[train_cols]

x_train = x_train.drop(columns=['Cabin', 'Emabrked'], errors='ignore')
x_test = x_test.drop(columns=['Cabin', 'Emabrked'], errors='ignore')

model=LogisticRegression(max_iter=1000)
model.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

In [154]:
#Evaluate Model
from sklearn.metrics import accuracy_score

x_test = x_test.fillna(0)

y_pred=model.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.8097014925373134

In [147]:
#Confusion Matrix
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
cm

array([[102,   3],
       [ 55,  19]])

True Positives
55
False Positives
15
False Negatives
19
True Negatives
90


In [148]:
#Classification Report
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.65      0.97      0.78       105
           1       0.86      0.26      0.40        74

    accuracy                           0.68       179
   macro avg       0.76      0.61      0.59       179
weighted avg       0.74      0.68      0.62       179



Why accuracy alone is not enough?
Precision tells us: out of all passengers predicted as “survived”, how many actually survived.
Recall tells us: out of all passengers who actually survived, how many we correctly identified.

In [149]:
#Baseline Comparison
y.value_counts(normalize=True)

Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64

If I predict “NOT survived” for everyone, what accuracy would I get?

~0.384

Day-10 Summary must include:

Learned confusion matrix

Learned precision vs recall

Why 1.0 accuracy is dangerous

What baseline accuracy means

Accuracy when test_size  is 0.2=0.6759776536312849
Accuracy when test_size is 0.3=0.8097014925373134

1)Increasing the test size from 0.2 to 0.3 changed the accuracy, showing that model performance depends on how data is split

2)Accuracy alone is misleading because it does not distinguish between false positives and false negatives and can hide poor performance on minority classes.

3)A false positive in the Titanic dataset means the model predicted a passenger survived when they actually did not.

4)Changing the test data proportion changes accuracy because the distribution and difficulty of test samples also change

Day10: Updated evaluation analysis and corrected accuracy interpretation
